File: JAVA/LongestSubarraySumK.java
import java.util.HashMap;

public class LongestSubarraySumK {
    public static int longestSubarraySumK(int[] nums, int k) {
        HashMap<Integer, Integer> prefixSumIndex = new HashMap<>();
        int sum = 0, maxLength = 0;

        for (int i = 0; i < nums.length; i++) {
            sum += nums[i];

            if (sum == k) {
                maxLength = i + 1;
            }

            if (!prefixSumIndex.containsKey(sum)) {
                prefixSumIndex.put(sum, i);
            }

            if (prefixSumIndex.containsKey(sum - k)) {
                maxLength = Math.max(maxLength, i - prefixSumIndex.get(sum - k));
            }
        }

        return maxLength;
    }

    public static void main(String[] args) {
        int[] nums1 = {10, 5, 2, 7, 1, 9};
        int k1 = 15;
        System.out.println(longestSubarraySumK(nums1, k1)); // Output: 4

        int[] nums2 = {-3, 2, 1};
        int k2 = 6;
        System.out.println(longestSubarraySumK(nums2, k2)); // Output: 0
    }
}

File: Python/LongestSubarraySumK.py
from typing import List

def longest_subarray_sum_k(nums: List[int], k: int) -> int:
    prefix_sum_index = {}
    total = 0
    max_len = 0

    for i, num in enumerate(nums):
        total += num

        if total == k:
            max_len = i + 1

        if total not in prefix_sum_index:
            prefix_sum_index[total] = i

        if total - k in prefix_sum_index:
            max_len = max(max_len, i - prefix_sum_index[total - k])

    return max_len

# Test cases
nums1 = [10, 5, 2, 7, 1, 9]
k1 = 15
print(longest_subarray_sum_k(nums1, k1))  # Output: 4

nums2 = [-3, 2, 1]
k2 = 6
print(longest_subarray_sum_k(nums2, k2))  # Output: 0
